airbyte-ci/connectors/base_images/base_images/registries.py

```diff
@@ -0,0 +1,237 @@
+class PublishedBaseImage(PublishedImage):
+    @property
+    def version(self) -> semver.VersionInfo:
+        return semver.VersionInfo.parse(self.tag)
+
+
+@dataclass
+class ChangelogEntry:
+    version: semver.VersionInfo
+    changelog_entry: str
+    dockerfile_example: str
+
+    def to_serializable_dict(self):
+        return {
+            "version": str(self.version),
+            "changelog_entry": self.changelog_entry,
+            "dockerfile_example": self.dockerfile_example,
+        }
+
+    @staticmethod
+    def from_dict(entry_dict: Dict):
+        return ChangelogEntry(
+            version=semver.VersionInfo.parse(entry_dict["version"]),
+            changelog_entry=entry_dict["changelog_entry"],
+            dockerfile_example=entry_dict["dockerfile_example"],
+        )
+
+
+@dataclass
+class RegistryEntry:
+    published_docker_image: Optional[PublishedBaseImage]
+    changelog_entry: Optional[ChangelogEntry]
+    version: semver.VersionInfo
+
+    @property
+    def published(self) -> bool:
+        return self.published_docker_image is not None
+
+
+class VersionRegistry:
+    def __init__(
+        self,
+        ConnectorBaseImageClass: Type[AirbyteConnectorBaseImage],
+        entries: List[RegistryEntry],
+    ) -> None:
+        self.ConnectorBaseImageClass: Type[AirbyteConnectorBaseImage] = ConnectorBaseImageClass
+        self._entries: List[RegistryEntry] = entries
+
+    @staticmethod
+    def get_changelog_dump_path(ConnectorBaseImageClass: Type[AirbyteConnectorBaseImage]) -> Path:
+        """Returns the path where the changelog is dumped to disk.
+
+        Args:
+            ConnectorBaseImageClass (Type[AirbyteConnectorBaseImage]): The base image version class bound to the registry.
+
+        Returns:
+            Path: The path where the changelog JSON is dumped to disk.
+        """
+        registries_dir = Path("generated/changelogs")
+        registries_dir.mkdir(exist_ok=True, parents=True)
+        return registries_dir / f'{ConnectorBaseImageClass.image_name.replace("-", "_").replace("/", "_")}.json'  # type: ignore
+
+    @property
+    def changelog_dump_path(self) -> Path:
+        """Returns the path where the changelog JSON is dumped to disk.
+
+        Returns:
+            Path: The path where the changelog JSON is dumped to disk.
+        """
+        return self.get_changelog_dump_path(self.ConnectorBaseImageClass)
+
+    @staticmethod
+    async def get_published_base_images(
+        ConnectorBaseImageClass: Type[AirbyteConnectorBaseImage], dagger_client: dagger.Client, docker_credentials: Tuple[str, str]
+    ) -> List[PublishedBaseImage]:
+        repository_address = f"{consts.REMOTE_REGISTRY}/{ConnectorBaseImageClass.image_name}"
+        dockerhub_username_secret = dagger_client.set_secret("DOCKER_HUB_USERNAME", docker_credentials[0])
+        dockerhub_username_password = dagger_client.set_secret("DOCKER_HUB_PASSWORD", docker_credentials[1])
+        crane_container = (
```

Comment:
❓given that crane is a utility and likely used in other ways in the future. Should this live in a docker utility file? so that we can call it like so

```py
crane = docker.crane_init(username=docker_credentials[0], password=docker_credentials[1])
ls_output = await crane.ls()
digest = await crane.digest([address])
```

or 

```py
crane = docker.crane_init(username=docker_credentials[0], password=docker_credentials[1])
ls_output = await crane.cmd(["ls"])
digest = await crane.cmd(["digest", address])
```

Reply:
I introduced `CraneClient` + `RemoteRegistry` classes in this direction, under `utils/docker.py`